### PR TITLE
rust-analyzer: init at unstable-2020-03-09

### PR DESCRIPTION
**Pull request by oxalica**
###### Motivation for this change
Add rust-analyzer, an experimental modular compiler frontend for the Rust language.
Also add the vscode extension.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by arcnmx** [comment]
It might be nice if we could split this up into an unwrapped and wrapped derivation. `rustcSrc` makes a fair dent in the closure size when it might just be overridden and unused (my typical use of it is to wrap it as part of a shell for a specific rust channel).

Also someone with ofborg permissions may want to initiate a darwin build. In the past I've had to include `buildInputs = optionals hostPlatform.isDarwin [ darwin.cf-private darwin.apple_sdk.frameworks.CoreServices ];` but haven't checked in a while if it's still necessary.

Finally, not an important point, but I wondered if we could/should build it with the `jemalloc` feature? I think features still can't be combined with cargo's `-p` flag yet though so :(

**Review by jonringer — COMMENTED** [review]
I would split this into two commits:

```
rust-analyzer: init at unstable-2020-01-13
vscode-extensions.rust-analyzer: init at unstable-2020-01-13
```

**Comment by oxalica** [comment]
@arcnmx 
That's great. But it's not easy to build two packages with features on using a single cargo command. I read the docs of rust-analyzer, and the binary `ra_cli` is development-only and seems just for benchmarks. To make it easy, I just build the language server `ra_lsp_server`.
Also, it builds on darwin now.

@jonringer 
Splitted.

**Comment by oxalica** [comment]
Updated to 2020-01-20

**Comment by oxalica** [comment]
- Rebased and updated to unstable-2020-01-29.
- Make the vscode extension to use `unstable-` prefixed version of rust-analyzer, since it is also unstable.
- Add maintainers.
- Some tiny fixes.

**Comment by oxalica** [comment]
Latest version 2020-02-10 is blocked by #79148, which is merged into `staging` but not in `master` yet.

**Comment by oxalica** [comment]
Rebased and bumped to 2020-02-11.
Also fixed update script.

**Comment by oxalica** [comment]
Bumped to 2020-02-25 and fix binary name.

**Comment by oxalica** [comment]
- Bump to 2020-03-02.
- Setup fake `rustup` in checkPhase to avoid patching test code.
- Fix update script.

**Review by lovesegfault — APPROVED** [review]
LGTM

**Comment by lovesegfault** [comment]
@jonringer ping

**Comment by oxalica** [comment]
- Rebased (hash of rustc changed)
- Bump to 2020-03-09
- Fix update script: ignore pre-releases.

**Comment by jonringer** [comment]
I know that's "best practice" for node packages to freeze their dependencies, but I just really don't like having 2200+ lines for a single package

@worldofpeace thoughts?

**Comment by oxalica** [comment]
@jonringer 
> ..., but I just really don't like having 2200+ lines for a single package

How about doing something like `rustPlatform.buildRustPackage` do?
We can use a `nodeSha256` or similar to lock dependencies (that is, `node_modules`), then simply use `npm install` to produce them with network, which is allowed in a fixed output derivation.

**Comment by jonringer** [comment]
not entirely sure how vendoring works, but I would be very okay with having a lockfile+vendor hash

@bhipple  for vendoring

**Comment by emilazy** [comment]
`buildRustCrate` is much more in line with the Nix(pkgs) model; the increase in verbosity is also an increase in flexibility and purity, and IIRC there's been some movement to try and remove uses of the fixed-output-derivation-using `buildRustPackage` from nixpkgs.

**Comment by oxalica** [comment]
@emilazy 
So you think we should keep lock files?
Well, I have no idea about how to avoid nixpkgs growing too fast.
And since we don't (and cannot) port all npm/rust packages, these dependency derivations are single-use and not able to override/customize like regular packages. We may not benefit much from these locks.

@jonringer 
There is already [an example](https://github.com/NixOS/nixpkgs/blob/e113100be0207ec2cfd54c39a069b823aa56f58e/pkgs/applications/version-management/sourcehut/fetchNodeModules.nix) about vendoring `node_modules`.
Maybe I can move it to top-level for easier using.

**Comment by emilazy** [comment]
>  So you think we should keep lock files?
> Well, I have no idea about how to avoid nixpkgs growing too fast.

Yes. There's no shortage of space in nixpkgs, but there's a severe shortage of pure, reproducible package builds in the world :)

**Comment by bhipple** [comment]
For languages like C/C++/Python/Haskell/JavaScript that are dynamically linked
or scripting languages, it makes sense to have all of the libraries in the
dependency graph represented as discrete Nix packages, since these all have to
be installed to the user's machine, and in that case it's nice to avoid
duplication.

For statically linked application binaries where the upstream community has a
strong bias towards vendoring via lockfiles, there are significant gains in
simplicity/consistency if we take advantage of that [1]. That said, there's a big
risk that the tool doing the vendoring itself will change how it produces the
vendor directories, resulting in us losing all of our fixed-output derivation
hashes.

My personal opinion is that the gain in simplicity and reduction of maintenance
cost within NixPkgs outweighs the cost, but this is a debatable opinion
[2]. Note that bloating the size of the NixPkgs git repo with generated files is
not a free action, because it is already 1GB+ and cloning it takes a long time,
particularly on slow connections. With builtins fetchers or CI, cloning it may
need to be done more often than just the first time a user installs Nix on their
workstation and starts hacking on the code, so it's not a one-time cost either.

One thing I will say though: these builds *are* pure and *are* bitwise
deterministic / reproducible [3]; it's just that the dependency graph is
course-grained, rather than fine-grained. For statically linked applications,
this has implications on build farm efficiency, but no implications for
end-users.

[1] For more information on these tradeoffs, this talk from NixCon gives an
    excellent overview: https://www.youtube.com/watch?v=nXDumHZI2zg
[2] https://github.com/NixOS/nix/issues/2270
[3] For Rust specifically, as part of https://github.com/NixOS/nixpkgs/issues/79975
    we have also installed poison pills into the build phase to ensure that the
    FOD `cargoDeps` exactly matches the FOD `src`.

**Comment by emilazy** [comment]
It's true that the whole git history is over a gigabyte, but most fetches are via channels or GitHub's revision archives, which are currently under 30 MiB compressed, so I personally don't think the amount of raw source text in nixpkgs is a major problem.

**Comment by oxalica** [comment]
@bhipple 
Thanks for explanation. Keeping lock files is also fine and that's what I do in this PR.

r? @emilazy @arcnmx

**Comment by bqv** [comment]
@jonringer ping

**Comment by lovesegfault** [comment]
cc. @worldofpeace

**Comment by lovesegfault** [comment]
@oxalica this may need another update to 2020-03-23.

**Comment by oxalica** [comment]
@lovesegfault 2020-03-23 requires rustc 1.42.0 (#82447), which is still in branch `staging`.

**Comment by lovesegfault** [comment]
@oxalica Oh, right, I forgot `staging` _still_ hasn't been merged. :(

**Comment by lovesegfault** [comment]
Pinging @jonringer @worldofpeace for a possible merge.

**Comment by lovesegfault** [comment]
@oxalica 1.42 is in master, you should be able to update this now I think.

Pinging @jonringer for a review.

**Comment by oxalica** [comment]
Do we really need to separate files and commits about `rust-analyzer` and `vscode-extensions.rust-analyzer`? They are always updated together.
Would it be better to move the extension derivation (with `node-packages.nix`) to `pkgs/development/tools/rust/rust-analyzer` and make a single update commit?

**Comment by bqv** [comment]
They're separate packages. It's fine for the update commits to always come in pairs, and there's no harm in more granularity

**Comment by oxalica** [comment]
@bqv Note that `vscode-extensions.rust-analyzer` references `rust-analyzer.src` but has separated `node-packages.nix`. So the extension may break when someone checkout the commit of `rust-analyzer`, due to outdated `node-packages.nix`.

**Comment by bqv** [comment]
Granted I'm not overly familiar with vsix instrumentation in nix, but perhaps in this scenario it would be better to have the extension as an output of the main package then? I do see your point, however. My other solution would be to break the dependency between the packages. The duplication isn't too much of an issue since the source would be reused (identical hash) anyway.

**Comment by oxalica** [comment]
@bqv 
> ... it would be better to have the extension as an output of the main package then?
The extension is a node package but `rust-analyzer` itself is a rust package. We cannot mix them together.

**Comment by bqv** [comment]
In that case I would suggest refer to the source twice, once in each derivation/commit (unless anyone thinks that's a bad idea)

File: pkgs/misc/vscode-extensions/rust-analyzer/default.nix
Comment (Mic92) [inline pkgs/misc/vscode-extensions/rust-analyzer/default.nix:13]: Can you move this to /home/joerg/git/nixpkgs/pkgs/development/node-packages/node-packages-v10.json ?

Comment (Mic92) [reply pkgs/misc/vscode-extensions/rust-analyzer/default.nix:13]: Your update script could also update the version in that json file.

File: pkgs/development/tools/rust/rust-analyzer/update.sh
Comment (Mic92) [inline pkgs/development/tools/rust/rust-analyzer/update.sh:44]: You can also use [nix-prefetch](https://github.com/msteen/nix-prefetch). This is how it can be used with rust: https://github.com/Mic92/nix-update/blob/master/nix_update/update.py#L53

File: pkgs/misc/vscode-extensions/rust-analyzer/default.nix
Comment (oxalica) [reply pkgs/misc/vscode-extensions/rust-analyzer/default.nix:13]: @Mic92 
I saw the syntax is like `{"npm2nix": "git://github.com/NixOS/npm2nix.git#5.12.0"}`.  But is there any way to specify **a sub-directory** of github repository as the package root in `node-packages-v*.json`? The `package.json` is [under `editors/code`](https://github.com/rust-analyzer/rust-analyzer/tree/master/editors/code).

Comment (Mic92) [reply pkgs/misc/vscode-extensions/rust-analyzer/default.nix]: It is not supported. It might be possible to implement this in future, but the current solution looks good to me!

**Comment by Mic92** [comment]
Thanks!

**Comment by asdf8dfafjk** [comment]
I like your wrapping approach and would prefer this package stays , since per project shell.nix could be different, and hence the argument to rust-analyzer, that said, its probably worth mentioning that rust-analyzer is now part of standard rust distribution. https://github.com/rust-lang/rust/pull/72978

**Comment by oxalica** [comment]
@asdf8dfafjk 
As a *compiler-independent* tool, I think we cannot benefit a lot from using dist from rust toolchain.

In nixpkgs, we have only stable rustc and cargo, bootstrapped from previous stable version.
That's quite a long time between two stable release of rust, but rust-analyzer could be more actively updated.

**Pull request merged**
